File: 0/Templates/CSharp/ProjectTemplates/NavigationAppTemplate/UI/PageHeader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace $safeprojectname$.UI
{
    /// <summary>
    /// The page header control with support for search.
    /// </summary>
    public class PageHeader
        : Control
    {
        /// <summary>
        /// Identifies the Frame dependency property.
        /// </summary>
        public static readonly DependencyProperty FrameProperty = DependencyProperty.Register("Frame", typeof(Frame), typeof(PageHeader), new PropertyMetadata(null, OnFrameChanged));
        /// <summary>
        /// Identifies the Icon dependency property.
        /// </summary>
        public static readonly DependencyProperty IconProperty = DependencyProperty.Register("Icon", typeof(string), typeof(PageHeader), null);
        /// <summary>
        /// Identifies the IsSearchBoxVisible dependency property.
        /// </summary>
        public static readonly DependencyProperty IsSearchBoxVisibleProperty = DependencyProperty.Register("IsSearchBoxVisible", typeof(bool), typeof(PageHeader), new PropertyMetadata(true));
        /// <summary>
        /// Identifies the SearchTerm dependency property.
        /// </summary>
        public static readonly DependencyProperty SearchTermProperty = DependencyProperty.Register("SearchTerm", typeof(string), typeof(PageHeader), null);
        /// <summary>
        /// Identifies the Title dependency property.
        /// </summary>
        public static readonly DependencyProperty TitleProperty = DependencyProperty.Register("Title", typeof(string), typeof(PageHeader), null);

        private Button iconButton;

        /// <summary>
        /// Initializes a new instance of the <see cref="PageHeader"/> control.
        /// </summary>
        public PageHeader()
        {
            this.DefaultStyleKey = typeof(PageHeader);
        }

        private static void OnFrameChanged(DependencyObject o, DependencyPropertyChangedEventArgs args)
        {
            ((PageHeader)o).OnFrameChanged();
        }

        private void OnFrameChanged()
        {
            UpdateIconButtonState();
        }

        /// <summary>
        /// Occurs when the control's template has been applied to this instance.
        /// </summary>
        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

            if (this.iconButton != null) {
                this.iconButton.Click -= OnIconButtonClick;
            }

            this.iconButton = GetTemplateChild("IconButton") as Button;
            if (this.iconButton != null) {
                this.iconButton.Click += OnIconButtonClick;
            }

            UpdateIconButtonState();
        }

        private void OnIconButtonClick(object sender, RoutedEventArgs e)
        {
            // attemp to navigate back to the first element in the frame's navigation stack
            while (this.Frame?.CanGoBack ?? false) {
                this.Frame.GoBack();
            }
        }

        private void UpdateIconButtonState()
        {
            if (this.iconButton == null) {
                return;
            }
            this.iconButton.IsEnabled = this.Frame?.CanGoBack ?? false;
        }

        /// <summary>
        /// Gets or sets the controlling frame.
        /// </summary>
        public Frame Frame
        {
            get { return (Frame)GetValue(FrameProperty); }
            set { SetValue(FrameProperty, value); }
        }

        /// <summary>
        /// Gets or sets the icon.
        /// </summary>
        public string Icon
        {
            get { return (string)GetValue(IconProperty); }
            set { SetValue(IconProperty, value); }
        }

        /// <summary>
        /// Gets or sets a value indicating whether the search box is visible.
        /// </summary>
        public bool IsSearchBoxVisible
        {
            get { return (bool)GetValue(IsSearchBoxVisibleProperty); }
            set { SetValue(IsSearchBoxVisibleProperty, value); }
        }

        /// <summary>
        /// Gets or sets the search term.
        /// </summary>
        public string SearchTerm
        {
            get { return (string)GetValue(SearchTermProperty); }
            set { SetValue(SearchTermProperty, value); }
        }

        /// <summary>
        /// Gets or sets the title.
        /// </summary>
        public string Title
        {
            get { return (string)GetValue(TitleProperty); }
            set { SetValue(TitleProperty, value); }
        }
    }
}

File: 0/Templates/VisualBasic/ProjectTemplates/NavigationAppVBTemplate/UI/PageHeader.vb
Namespace UI
    ''' <summary>
    ''' The page header control with support for search.
    ''' </summary>
    Public Class PageHeader
        Inherits Control

        Public Shared ReadOnly FrameProperty As DependencyProperty = DependencyProperty.Register("Frame", GetType(Frame), GetType(PageHeader), New PropertyMetadata(Nothing, AddressOf OnFrameChanged))
        Public Shared ReadOnly IconProperty As DependencyProperty = DependencyProperty.Register("Icon", GetType(String), GetType(PageHeader), Nothing)
        Public Shared ReadOnly IsSearchBoxVisibleProperty As DependencyProperty = DependencyProperty.Register("IsSearchBoxVisible", GetType(Boolean), GetType(PageHeader), New PropertyMetadata(True))
        Public Shared ReadOnly SearchTermProperty As DependencyProperty = DependencyProperty.Register("SearchTerm", GetType(String), GetType(PageHeader), Nothing)
        Public Shared ReadOnly TitleProperty As DependencyProperty = DependencyProperty.Register("Title", GetType(String), GetType(PageHeader), Nothing)

        Private _iconButton As Button

        Public Sub New()
            DefaultStyleKey = GetType(PageHeader)
        End Sub

        Private Shared Sub OnFrameChanged(o As DependencyObject, args As DependencyPropertyChangedEventArgs)
            TryCast(o, PageHeader).OnFrameChanged()
        End Sub

        Private Sub OnFrameChanged()
            UpdateIconButtonState()
        End Sub

        Protected Overrides Sub OnApplyTemplate()
            MyBase.OnApplyTemplate()

            If Not _iconButton Is Nothing Then
                RemoveHandler _iconButton.Click, AddressOf OnIconButtonClick
            End If

            _iconButton = TryCast(GetTemplateChild("IconButton"), Button)
            If Not _iconButton Is Nothing Then
                AddHandler _iconButton.Click, AddressOf OnIconButtonClick
            End If
        End Sub

        Private Sub OnIconButtonClick(sender As Object, e As RoutedEventArgs)
            If Frame Is Nothing Then Return

            ' attemp to navigate back to the first element in the frame's navigation stack
            While Frame.CanGoBack
                Frame.GoBack()
            End While
        End Sub

        Private Sub UpdateIconButtonState()
            If _iconButton Is Nothing Or Frame Is Nothing Then Return

            _iconButton.IsEnabled = Frame.CanGoBack
        End Sub

        Public Property Frame As Frame
            Get
                Return GetValue(FrameProperty)
            End Get
            Set(value As Frame)
                SetValue(FrameProperty, value)
            End Set
        End Property

        Public Property IsSearchBoxVisible As Boolean
            Get
                Return GetValue(IsSearchBoxVisibleProperty)
            End Get
            Set(value As Boolean)
                SetValue(IsSearchBoxVisibleProperty, value)
            End Set
        End Property

        Public Property Icon As String
            Get
                Return GetValue(IconProperty)
            End Get
            Set(value As String)
                SetValue(IconProperty, value)
            End Set
        End Property

        Public Property SearchTerm As String
            Get
                Return GetValue(SearchTermProperty)
            End Get
            Set(value As String)
                SetValue(SearchTermProperty, value)
            End Set
        End Property

        Public Property Title As String
            Get
                Return GetValue(TitleProperty)
            End Get
            Set(value As String)
                SetValue(TitleProperty, value)
            End Set
        End Property
    End Class
End Namespace

